services/placement/filter/mirrored.go

```diff
@@ -0,0 +1,435 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package filter
+
+import (
+	"container/heap"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+)
+
+var (
+	errNoValidMirrorInstance = errors.New("no valid instance for mirror placement in the candidate list")
+)
+
+type mirroredFilter struct {
+	opts services.PlacementOptions
+}
+
+func newMirroredFilter(opts services.PlacementOptions) placement.InstanceFilter {
+	return &mirroredFilter{opts: opts}
+}
+
+func (f *mirroredFilter) FilterInitialInstances(
+	candidates []services.PlacementInstance,
+	rf int,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(placement.NewPlacement(), candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	weightToHostMap, err := hostsGroupedByWeight(candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	var pairs = make([][]services.PlacementInstance, 0, len(candidates))
+	for _, hosts := range weightToHostMap {
+		pairedHosts := pairHosts(hosts, rf)
+		if len(pairedHosts) == 0 {
+			continue
+		}
+
+		pairedInstances, err := pairInstances(pairedHosts)
+		if err != nil {
+			return nil, err
+		}
+
+		pairs = append(pairs, pairedInstances...)
+	}
+
+	if len(pairs) == 0 {
+		return nil, errNoValidMirrorInstance
+	}
+
+	return assignShardsToPairedInstances(pairs, placement.NewPlacement()), nil
+}
+
+func (f *mirroredFilter) FilterAddingInstances(
+	candidates []services.PlacementInstance,
+	p services.Placement,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(p, candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	weightToHostMap, err := hostsGroupedByWeight(candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	var pairs = make([][]services.PlacementInstance, 0, len(candidates))
+	for _, hosts := range weightToHostMap {
+		pairedHosts := pairHosts(hosts, p.ReplicaFactor())
+		if len(pairedHosts) == 0 {
+			continue
+		}
+
+		pairs, err = pairInstances(pairedHosts[:1])
```

Comment:
Right I should've clarified, I meant when we got an error here when grouping instances, do we wanna continue trying other groups too instead of bailing early, given that there may be eligible instances in other groups that could satisfy the request?

Also is there any logic that prefers instances with larger or smaller weights?